logging_middleware.go

```diff
@@ -1,26 +1,52 @@
 package handlers
 
 import (
-	"log"
+	"context"
 	"net/http"
 	"time"
 
+	"github.com/Sirupsen/logrus"
 	"github.com/codegangsta/negroni"
 )
 
 type LoggingMiddleware struct {
-	logger *log.Logger
+	logger logrus.FieldLogger
 }
 
-func NewLoggingMiddleware(logger *log.Logger) Middleware {
+func NewLoggingMiddleware(logger *logrus.Logger) Middleware {
 	m := &LoggingMiddleware{logger}
 	return m
 }
 
 func (l *LoggingMiddleware) Apply(next HandlerFunc) HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+		logger := l.logger
 		before := time.Now()
-		l.logger.Printf("%v %v %v - %v", before, r.Method, r.URL, r.RemoteAddr)
+
+		id, ok := r.Context().Value("request_id").(string)
+		if ok {
+			logger = logger.WithField("request_id", id)
+		}
+
+		r = r.WithContext(context.WithValue(r.Context(), "logger", logger))
+
+		fields := logrus.Fields{
+			"method":     r.Method,
+			"path":       r.URL.String(),
+			"host":       r.Host,
+			"from":       r.RemoteAddr,
+			"protocol":   r.Proto,
+			"referer":    r.Referer(),
+			"user_agent": r.UserAgent(),
+		}
+		for k, v := range fields {
+			if len(v.(string)) == 0 {
+				delete(fields, k)
+			}
+		}
+		logger = logger.WithFields(fields)
+		logger.Info()
```

Comment:
Can you explain this line?

Reply:
The `.Info()` line?

Reply:
Once we have an Entry, it has to be written on stdout/stderr/where the logger writes